pkg/cluster-runtime/installer.go

```diff
@@ -118,71 +124,126 @@ func getWorkerIPList(infraDriver infradriver.InfraDriver) []net.IP {
 	return workers
 }
 
-func (i *Installer) Install() (registry.Driver, runtime.Driver, error) {
-	master0 := i.infraDriver.GetHostIPListByRole(common.MASTER)[0]
-	masters := i.infraDriver.GetHostIPListByRole(common.MASTER)
-	workers := getWorkerIPList(i.infraDriver)
-	all := append(masters, workers...)
+func (i *Installer) Install() error {
+	var (
+		masters    = i.infraDriver.GetHostIPListByRole(common.MASTER)
+		master0    = masters[0]
+		workers    = getWorkerIPList(i.infraDriver)
+		all        = append(masters, workers...)
+		image      = i.infraDriver.GetClusterImageName()
+		launchCmds = i.infraDriver.GetClusterLaunchCmds()
+	)
 
 	// distribute rootfs
 	if err := i.Distributor.DistributeRootfs(all, i.infraDriver.GetClusterRootfsPath()); err != nil {
-		return nil, nil, err
+		return err
 	}
 
 	if err := i.runClusterHook(master0, PreInstallCluster); err != nil {
-		return nil, nil, err
+		return err
 	}
 
 	if err := i.runHostHook(PreInitHost, all); err != nil {
-		return nil, nil, err
+		return err
 	}
 
+	extension, err := i.ImageEngine.GetSealerImageExtension(&common2.GetImageAnnoOptions{ImageNameOrID: image})
+	if err != nil {
+		return fmt.Errorf("failed to get ClusterImage: %s", err)
+	}
+
+	if extension.Type == v12.AppInstaller {
+		var cmds []string
+		if len(launchCmds) > 0 {
+			cmds = launchCmds
+		} else {
+			cmds = extension.Launch.Cmds
+		}
+
+		err = i.installKubeCluster(all)
+		if err != nil {
+			return fmt.Errorf("failed to install cluster: %s", err)
+		}
+
+		err = i.installApp(master0, cmds)
+		if err != nil {
+			return fmt.Errorf("failed to install application: %s", err)
+		}
+	} else {
```

Comment:
Where will the `CMDS` defined in the old image be executed?

Reply:
got it from image extension.Launch.Cmds